shuup/core/api/basket.py

```diff
@@ -297,42 +296,71 @@ def _handle_cmd(self, request, command, kwargs):
         response = cmd_handler(**cmd_kwargs)
         return cmd_dispatcher.postprocess_response(command, cmd_kwargs, response)
 
-    @schema_serializer_class(ProductAddBasketSerializer)
-    @detail_route(methods=['post'])
-    def add(self, request, *args, **kwargs):
-        """
-        Adds a product to the basket
-        """
-        self.process_request()
-        if "shop_product" in request.data:
-            serializer = ShopProductAddBasketSerializer(data=request.data, context={"shop": request.shop})
+    def add_product(self, data):
+        if "shop_product" in data:
+            serializer = ShopProductAddBasketSerializer(data=data, context={"shop": self.request.shop})
         else:
-            serializer = ProductAddBasketSerializer(data=request.data, context={"shop": request.shop})
+            serializer = ProductAddBasketSerializer(data=data, context={"shop": self.request.shop})
 
-        if serializer.is_valid():
+        if serializer.is_valid(raise_exception=True):
             cmd_kwargs = {
-                "request": request._request,
-                "basket": request._request.basket,
+                "request": self.request._request,
+                "basket": self.request._request.basket,
                 "shop_id": serializer.data.get("shop") or serializer.validated_data["shop"].pk,
                 "product_id": serializer.data.get("product") or serializer.validated_data["product"].pk,
                 "quantity": serializer.validated_data.get("quantity", 1),
                 "supplier_id": serializer.validated_data.get("supplier")
             }
-            # we call `add` directly, assuming the user will handle variations
-            # as he can know all product variations easily through product API
+            self._handle_cmd(self.request._request, "add", cmd_kwargs)
+            self.request.basket.save()
+
+    @detail_route(methods=['post'])
+    def add(self, request, *args, **kwargs):
```

Comment:
put the `@schema_serializer_class(ProductAddBasketSerializer)` here. it is important when generating the documentation, otherwise the docs will show the wrong serializer.